Sources_v3/StreamChat/Database/DTOs/AttachmentDTO.swift

```
@@ -0,0 +1,206 @@
+        dto.author = payload.author
+        dto.text = payload.text
+        dto.type = payload.type.rawValue ?? "unknown"
+        dto.actions = try JSONEncoder.stream.encode(payload.actions)
+        dto.url = payload.url
+        dto.imageURL = payload.imageURL
+        dto.file = try JSONEncoder.stream.encode(payload.file)
+        dto.extraData = try JSONEncoder.default.encode(payload.extraData)
+        
+        if let messageDTO = MessageDTO.load(id: messageId, context: self) {
+            dto.message = messageDTO
+            messageDTO.attachments.insert(dto)
+        }
+        
+        if let channelDTO = ChannelDTO.load(cid: cid, context: self) {
+            dto.channel = channelDTO
+            channelDTO.attachments.insert(dto)
+        }
+        
+        return dto
+    }
+    
+    func saveAttachment<ExtraData: AttachmentExtraData>(
+        attachment: _ChatMessageAttachment<ExtraData>,
+        messageId: MessageId,
+        cid: ChannelId
+    ) throws -> AttachmentDTO {
+        let dto = AttachmentDTO.loadOrCreate(queryHash: attachment.hash, context: self)
+        dto.attachmentHash = attachment.hash
+        dto.title = attachment.title
+        dto.author = attachment.author
+        dto.text = attachment.text
+        dto.type = attachment.type.rawValue ?? "unknown"
+        dto.actions = try JSONEncoder.stream.encode(attachment.actions)
+        dto.url = attachment.url
+        dto.imageURL = attachment.imageURL
+        dto.file = try JSONEncoder.stream.encode(attachment.file)
+        dto.extraData = try JSONEncoder.default.encode(attachment.extraData)
+        
+        if let messageDTO = MessageDTO.load(id: messageId, context: self) {
+            dto.message = messageDTO
+            messageDTO.attachments.insert(dto)
+        }
+        
+        if let channelDTO = ChannelDTO.load(cid: cid, context: self) {
+            dto.channel = channelDTO
+            channelDTO.attachments.insert(dto)
+        }
```

We should throw an error here if we don't find the message or channel DTOs.